### PR TITLE
Wrench Reftests

### PR DESCRIPTION
**Pull request by jrmuizel**
This is my work-in-progress that adds reftesting functionality to wrench

---
This change is [Reviewable](https://reviewable.io/reviews/servo/webrender/695)

### CODE REVIEW

**Comment by glennw** [comment]
@jrmuizel Cool! Did you want this stuff reviewed / merged now, or wait until the [WIP] is removed?

**Comment by jrmuizel** [comment]
I'll put a reviewable version up tomorrow.

**Comment by kvark** [comment]
Can we have the `mask.yaml`, `mask.png`, and the reftest list in different folder than `src`?

File: wrench/src/reftest.rs [hunk 1]
```diff
@@ -0,0 +1,78 @@
...
+        let s = &l[0..l.find("#").unwrap_or(l.len())];
+        let s = s.trim();
+        if l.len() == 0 {
+            continue;
+        }
+        let mut items = s.split_whitespace();
+        let kind = match items.next() {
+            Some("==") => ReftestOp::Equal,
+            Some("!=") => ReftestOp::NotEqual,
+            _ => panic!()
```
Comment (kvark) [inline wrench/src/reftest.rs]: nit: maybe just `continue` with a warning instead

File: wrench/src/reftest.rs [hunk 2]
```diff
@@ -0,0 +1,78 @@
...
+}
+
+fn render_yaml(wrench: &mut Wrench, window: &mut WindowWrapper, filename: &str) -> Vec<u8>
+{
+    let mut reader = YamlFrameReader::new(Path::new(filename));
+    reader.do_frame(wrench);
+    wrench.render();
+    window.swap_buffers();
+    let size = window.get_inner_size();
+    gl::read_pixels(0, 0,
```
Comment (kvark) [inline wrench/src/reftest.rs]: uh oh, we need to hide it behind WR's device abstraction

File: wrench/src/reftest.rs [hunk 3]
```diff
@@ -0,0 +1,78 @@
...
+                    size.0 as gl::GLsizei,
+                    size.1 as gl::GLsizei,
+                    gl::RGB, gl::UNSIGNED_BYTE)
```
Comment (kvark) [inline wrench/src/reftest.rs]: Can we use `RGBA` instead in order to avoid internal format conversion?

File: wrench/src/reftest.rs [hunk 4]
```diff
@@ -0,0 +1,78 @@
...
+}
+
+pub fn run_reftests(wrench: &mut Wrench, window: &mut WindowWrapper, filename: &str)
+{
+    parse_reftests(filename, |t: Reftest|
+                   {
```
Comment (kvark) [inline wrench/src/reftest.rs]: formatting

**Comment by jrmuizel** [comment]
This should be ready for review now.

File: wrench/src/reftest.rs [hunk 5]
```diff
@@ -0,0 +1,108 @@
...
+pub fn parse_reftests<F>(filename: &str, mut runner: F)
+    where F: FnMut(Reftest)
+{
+    let manifest = Path::new(filename);
+    let dir = manifest.parent().unwrap();
+    let f = File::open(manifest).unwrap();
+    let file = BufReader::new(&f);
+    for line in file.lines() {
+        let l = line.unwrap();
+        if l.starts_with("#") {
```
Comment (kvark) [inline wrench/src/reftest.rs]: not really needed here, the next `continue` would do

**Comment by kvark** [comment]
Thanks @jrmuizel !
It appears that we are already using `gl::read_pixels` in `wrench`, so a fix to both this and the older case could be a follow-up.
@glennw leaving for you to either review or merge with r=me

File: wrench/src/main.rs [hunk 6]
```diff
@@ -54,6 +54,9 @@ use yaml_frame_reader::YamlFrameReader;
 
 mod yaml_frame_writer;
 mod json_frame_writer;
+mod scene;
```
Comment (kvark) [inline wrench/src/main.rs]: this looks like a leftover, and it fails to build

File: wrench/src/reftest.rs [hunk 7]
```diff
@@ -0,0 +1,108 @@
...
+            continue;
+        }
+
+        // strip the comments
+        let s = &l[0..l.find("#").unwrap_or(l.len())];
+        let s = s.trim();
+        if l.len() == 0 {
```
Comment (kvark) [inline wrench/src/reftest.rs]: that is comparing `l.len()` and not `s.len()`? I misread it at first.

File: wrench/src/reftest.rs (hunk 5 above)
Comment (kvark) [inline wrench/src/reftest.rs]: actually, the next `continue` is checking for `l.len()==0` so it's different

File: wrench/src/reftest.rs [hunk 8]
```diff
@@ -0,0 +1,108 @@
...
+    let file = BufReader::new(&f);
+    for line in file.lines() {
+        let l = line.unwrap();
+        if l.starts_with("#") {
+            continue;
+        }
+
+        // strip the comments
+        let s = &l[0..l.find("#").unwrap_or(l.len())];
+        let s = s.trim();
```
Comment (kvark) [inline wrench/src/reftest.rs]: a simpler way, possibly, is: `let s = l.split('#').next().unwrap().trim()`

**Comment by glennw** [comment]
This is awesome, thanks!

**Pull request merged**